File: core/src/main/java/io/github/mathter/memifydb/core/util/ByteArray.java
package io.github.mathter.memifydb.core.util;

import java.io.EOFException;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.lang.invoke.MethodHandles;
import java.lang.invoke.VarHandle;
import java.nio.ByteOrder;

public class ByteArray {
    private static final VarHandle SHORT = create(short[].class);

    private static final VarHandle CHAR = create(char[].class);

    private static final VarHandle INT = create(int[].class);

    private static final VarHandle LONG = create(long[].class);

    private static final VarHandle FLOAT = create(float[].class);

    private static final VarHandle DOUBLE = create(double[].class);

    private static VarHandle create(Class<?> viewArrayClass) {
        return MethodHandles.byteArrayViewVarHandle(viewArrayClass, ByteOrder.BIG_ENDIAN);
    }

    public static byte getByteRaw(byte[] buf, int offset) {
        return buf[offset];
    }

    public static void setByteRaw(byte[] buf, int offset, byte value) {
        buf[offset] = value;
    }

    public static short getShortRaw(byte[] buf, int offset) {
        return (short) SHORT.get(buf, offset);
    }

    public static void setShortRaw(byte[] buf, int offset, short value) {
        SHORT.set(buf, offset, value);
    }

    public static char getCharRaw(byte[] buf, int offset) {
        return (char) CHAR.get(buf, offset);
    }

    public static void setCharRaw(byte[] buf, int offset, char value) {
        CHAR.set(buf, offset, value);
    }

    public static int getIntRaw(byte[] buf, int offset) {
        return (int) INT.get(buf, offset);
    }

    public static void setIntRaw(byte[] buf, int offset, int value) {
        INT.set(buf, offset, value);
    }

    public static long getLongRaw(byte[] buf, int offset) {
        return (long) LONG.get(buf, offset);
    }

    public static void setLongRaw(byte[] buf, int offset, long value) {
        LONG.set(buf, offset, value);
    }

    public static float getFloat(byte[] buf, int offset) {
        return (float) Float.intBitsToFloat((int) INT.get(buf, offset));
    }

    public static void setFloat(byte[] buf, int offset, float value) {
        INT.set(buf, offset, Float.floatToIntBits(value));
    }

    public static float getFloatRaw(byte[] buf, int offset) {
        return (float) FLOAT.get(buf, offset);
    }

    public static void setFloatRaw(byte[] buf, int offset, float value) {
        FLOAT.set(buf, offset, value);
    }

    public static double getDouble(byte[] buf, int offset) {
        return Double.longBitsToDouble((long) LONG.get(buf, offset));
    }

    public static void setDouble(byte[] buf, int offset, double value) {
        LONG.set(buf, offset, Double.doubleToLongBits(value));
    }

    public static double getDoubleRaw(byte[] buf, int offset) {
        return (double) DOUBLE.get(buf, offset);
    }

    public static void setDoubleRaw(byte[] buf, int offset, double value) {
        DOUBLE.set(buf, offset, value);
    }

    public static byte readByteRaw(InputStream is) throws IOException {
        int result = is.read();
        if (result == -1) {
            throw new EOFException("Unexpected end of stream");
        }
        return (byte) result;
    }

    public static void writeByteRaw(OutputStream os, byte value) throws IOException {
        os.write(value);
    }

    public static short readShortRaw(InputStream is) throws IOException {
        final byte[] buf = new byte[Short.BYTES];
        final int read = is.readNBytes(buf, 0, buf.length);
        if (read != buf.length) {
            throw new EOFException("Expected " + buf.length + " bytes, got " + read);
        }
        return (short) SHORT.get(buf, 0);
    }

    public static void writeShortRaw(OutputStream os, short value) throws IOException {
        final byte[] buf = new byte[Short.BYTES];
        SHORT.set(buf, 0, value);

        os.write(buf);
    }

    public static char readCharRaw(InputStream is) throws IOException {
        final byte[] buf = new byte[Character.BYTES];
        final int read = is.readNBytes(buf, 0, buf.length);
        if (read != buf.length) {
            throw new EOFException("Expected " + buf.length + " bytes, got " + read);
        }
        return (char) CHAR.get(buf, 0);
    }

    public static void writeCharRaw(OutputStream os, char value) throws IOException {
        final byte[] buf = new byte[Character.BYTES];
        CHAR.set(buf, 0, value);

        os.write(buf);
    }

    public static int readIntRaw(InputStream is) throws IOException {
        final byte[] buf = new byte[Integer.BYTES];
        final int read = is.readNBytes(buf, 0, buf.length);
        if (read != buf.length) {
            throw new EOFException("Expected " + buf.length + " bytes, got " + read);
        }
        return (int) INT.get(buf, 0);
    }

    public static void writeIntRaw(OutputStream os, int value) throws IOException {
        final byte[] buf = new byte[Integer.BYTES];
        INT.set(buf, 0, value);

        os.write(buf);
    }

    public static long readLongRaw(InputStream is) throws IOException {
        final byte[] buf = new byte[Long.BYTES];
        final int read = is.readNBytes(buf, 0, buf.length);
        if (read != buf.length) {
            throw new EOFException("Expected " + buf.length + " bytes, got " + read);
        }
        return (long) LONG.get(buf, 0);
    }

    public static void writeLongRaw(OutputStream os, long value) throws IOException {
        final byte[] buf = new byte[Long.BYTES];
        LONG.set(buf, 0, value);

        os.write(buf);
    }

    public static float readFloat(InputStream is) throws IOException {
        final byte[] buf = new byte[Integer.BYTES];
        final int read = is.readNBytes(buf, 0, buf.length);
        if (read != buf.length) {
            throw new EOFException("Expected " + buf.length + " bytes, got " + read);
        }
        return Float.intBitsToFloat((int) INT.get(buf, 0));
    }

    public static void writeFloat(OutputStream os, float value) throws IOException {
        final byte[] buf = new byte[Integer.BYTES];
        INT.set(buf, 0, Float.floatToRawIntBits(value));

        os.write(buf);
    }

    public static float readFloatRaw(InputStream is) throws IOException {
        final byte[] buf = new byte[Float.BYTES];
        final int read = is.readNBytes(buf, 0, buf.length);
        if (read != buf.length) {
            throw new EOFException("Expected " + buf.length + " bytes, got " + read);
        }
        return (float) FLOAT.get(buf, 0);
    }

    public static void writeFloatRaw(OutputStream os, float value) throws IOException {
        final byte[] buf = new byte[Float.BYTES];
        FLOAT.set(buf, 0, value);

        os.write(buf);
    }

    public static double readDouble(InputStream is) throws IOException {
        final byte[] buf = new byte[Long.BYTES];
        final int read = is.readNBytes(buf, 0, buf.length);
        if (read != buf.length) {
            throw new EOFException("Expected " + buf.length + " bytes, got " + read);
        }
        return Double.longBitsToDouble((long) LONG.get(buf, 0));
    }

    public static void writeDouble(OutputStream os, double value) throws IOException {
        final byte[] buf = new byte[Long.BYTES];
        LONG.set(buf, 0, Double.doubleToLongBits(value));

        os.write(buf);
    }

    public static double readDoubleRaw(InputStream is) throws IOException {
        final byte[] buf = new byte[Double.BYTES];
        final int read = is.readNBytes(buf, 0, buf.length);
        if (read != buf.length) {
            throw new EOFException("Expected " + buf.length + " bytes, got " + read);
        }
        return (double) DOUBLE.get(buf, 0);
    }

    public static void writeDoubleRaw(OutputStream os, double value) throws IOException {
        final byte[] buf = new byte[Double.BYTES];
        DOUBLE.set(buf, 0, value);

        os.write(buf);
    }
}

File: common/src/main/scala/io/github/mathter/memifydb/common/util/ByteArray.scala
package io.github.mathter.memifydb.common.util

import java.io.{EOFException, IOException, InputStream, OutputStream}
import java.lang
import java.lang.invoke.{MethodHandles, VarHandle}
import java.nio.ByteOrder

/**
 * Copyright 2026 Alexander Kashirsky (mathter)
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
object ByteArray {
  private val SHORT = create(classOf[Array[Short]])

  private val CHAR = create(classOf[Array[Char]])

  private val INT = create(classOf[Array[Int]])

  private val LONG = create(classOf[Array[Long]])

  private val FLOAT = create(classOf[Array[Float]])

  private val DOUBLE = create(classOf[Array[Double]])

  private def create(viewArrayClass: Class[_]) = MethodHandles.byteArrayViewVarHandle(viewArrayClass, ByteOrder.BIG_ENDIAN)

  inline def getByteRaw(buf: Array[Byte], offset: Int): Byte = buf(offset)

  inline def setByteRaw(buf: Array[Byte], offset: Int, value: Byte): Unit = {
    buf(offset) = value
  }

  inline def getShortRaw(buf: Array[Byte], offset: Int): Short = SHORT.get(buf, offset).asInstanceOf[Short]

  inline def setShortRaw(buf: Array[Byte], offset: Int, value: Short): Unit = {
    SHORT.set(buf, offset, value)
  }

  inline def getCharRaw(buf: Array[Byte], offset: Int): Char = CHAR.get(buf, offset).asInstanceOf[Char]

  inline def setCharRaw(buf: Array[Byte], offset: Int, value: Char): Unit = {
    CHAR.set(buf, offset, value)
  }

  inline def getIntRaw(buf: Array[Byte], offset: Int): Int = INT.get(buf, offset).asInstanceOf[Int]

  inline def setIntRaw(buf: Array[Byte], offset: Int, value: Int): Unit = {
    INT.set(buf, offset, value)
  }

  inline def getLongRaw(buf: Array[Byte], offset: Int): Long = LONG.get(buf, offset).asInstanceOf[Long]

  inline def setLongRaw(buf: Array[Byte], offset: Int, value: Long): Unit = {
    LONG.set(buf, offset, value)
  }

  inline def getFloat(buf: Array[Byte], offset: Int): Float = lang.Float.intBitsToFloat(INT.get(buf, offset).asInstanceOf[Int]).toFloat

  inline def setFloat(buf: Array[Byte], offset: Int, value: Float): Unit = {
    INT.set(buf, offset, lang.Float.floatToIntBits(value))
  }

  inline def getFloatRaw(buf: Array[Byte], offset: Int): Float = FLOAT.get(buf, offset).asInstanceOf[Float]

  inline def setFloatRaw(buf: Array[Byte], offset: Int, value: Float): Unit = {
    FLOAT.set(buf, offset, value)
  }

  inline def getDouble(buf: Array[Byte], offset: Int): Double = lang.Double.longBitsToDouble(LONG.get(buf, offset).asInstanceOf[Long])

  inline def setDouble(buf: Array[Byte], offset: Int, value: Double): Unit = {
    LONG.set(buf, offset, lang.Double.doubleToLongBits(value))
  }

  inline def getDoubleRaw(buf: Array[Byte], offset: Int): Double = DOUBLE.get(buf, offset).asInstanceOf[Double]

  inline def setDoubleRaw(buf: Array[Byte], offset: Int, value: Double): Unit = {
    DOUBLE.set(buf, offset, value)
  }

  @throws[IOException]
  inline def readByteRaw(is: InputStream): Byte = {
    val result = is.read
    if (result == -1) throw new EOFException("Unexpected end of stream")
    result.toByte
  }

  @throws[IOException]
  inline def writeByteRaw(os: OutputStream, value: Byte): Unit = {
    os.write(value.asInstanceOf[Int])
  }

  @throws[IOException]
  inline def readShortRaw(is: InputStream): Short = {
    val buf = new Array[Byte](lang.Short.BYTES)
    val read = is.readNBytes(buf, 0, buf.length)
    if (read != buf.length) throw new EOFException("Expected " + buf.length + " bytes, got " + read)
    SHORT.get(buf, 0).asInstanceOf[Short]
  }

  @throws[IOException]
  inline def writeShortRaw(os: OutputStream, value: Short): Unit = {
    val buf = new Array[Byte](lang.Short.BYTES)
    SHORT.set(buf, 0, value)
    os.write(buf)
  }

  @throws[IOException]
  inline def readCharRaw(is: InputStream): Char = {
    val buf = new Array[Byte](Character.BYTES)
    val read = is.readNBytes(buf, 0, buf.length)
    if (read != buf.length) throw new EOFException("Expected " + buf.length + " bytes, got " + read)
    CHAR.get(buf, 0).asInstanceOf[Char]
  }

  @throws[IOException]
  inline def writeCharRaw(os: OutputStream, value: Char): Unit = {
    val buf = new Array[Byte](Character.BYTES)
    CHAR.set(buf, 0, value)
    os.write(buf)
  }

  @throws[IOException]
  inline def readIntRaw(is: InputStream): Int = {
    val buf = new Array[Byte](Integer.BYTES)
    val read = is.readNBytes(buf, 0, buf.length)
    if (read != buf.length) throw new EOFException("Expected " + buf.length + " bytes, got " + read)
    INT.get(buf, 0).asInstanceOf[Int]
  }

  @throws[IOException]
  inline def writeIntRaw(os: OutputStream, value: Int): Unit = {
    val buf = new Array[Byte](Integer.BYTES)
    INT.set(buf, 0, value)
    os.write(buf)
  }

  @throws[IOException]
  inline def readLongRaw(is: InputStream): Long = {
    val buf = new Array[Byte](lang.Long.BYTES)
    val read = is.readNBytes(buf, 0, buf.length)
    if (read != buf.length) throw new EOFException("Expected " + buf.length + " bytes, got " + read)
    LONG.get(buf, 0).asInstanceOf[Long]
  }

  @throws[IOException]
  inline def writeLongRaw(os: OutputStream, value: Long): Unit = {
    val buf = new Array[Byte](lang.Long.BYTES)
    LONG.set(buf, 0, value)
    os.write(buf)
  }

  @throws[IOException]
  inline def readFloat(is: InputStream): Float = {
    val buf = new Array[Byte](Integer.BYTES)
    val read = is.readNBytes(buf, 0, buf.length)
    if (read != buf.length) throw new EOFException("Expected " + buf.length + " bytes, got " + read)
    lang.Float.intBitsToFloat(INT.get(buf, 0).asInstanceOf[Int])
  }

  @throws[IOException]
  inline def writeFloat(os: OutputStream, value: Float): Unit = {
    val buf = new Array[Byte](Integer.BYTES)
    INT.set(buf, 0, lang.Float.floatToRawIntBits(value))
    os.write(buf)
  }

  @throws[IOException]
  inline def readFloatRaw(is: InputStream): Float = {
    val buf = new Array[Byte](lang.Float.BYTES)
    val read = is.readNBytes(buf, 0, buf.length)
    if (read != buf.length) throw new EOFException("Expected " + buf.length + " bytes, got " + read)
    FLOAT.get(buf, 0).asInstanceOf[Float]
  }

  @throws[IOException]
  inline def writeFloatRaw(os: OutputStream, value: Float): Unit = {
    val buf = new Array[Byte](lang.Float.BYTES)
    FLOAT.set(buf, 0, value)
    os.write(buf)
  }

  @throws[IOException]
  inline def readDouble(is: InputStream): Double = {
    val buf = new Array[Byte](lang.Long.BYTES)
    val read = is.readNBytes(buf, 0, buf.length)
    if (read != buf.length) throw new EOFException("Expected " + buf.length + " bytes, got " + read)
    lang.Double.longBitsToDouble(LONG.get(buf, 0).asInstanceOf[Long])
  }

  @throws[IOException]
  inline def writeDouble(os: OutputStream, value: Double): Unit = {
    val buf = new Array[Byte](lang.Long.BYTES)
    LONG.set(buf, 0, lang.Double.doubleToLongBits(value))
    os.write(buf)
  }

  @throws[IOException]
  inline def readDoubleRaw(is: InputStream): Double = {
    val buf = new Array[Byte](lang.Double.BYTES)
    val read = is.readNBytes(buf, 0, buf.length)
    if (read != buf.length) throw new EOFException("Expected " + buf.length + " bytes, got " + read)
    DOUBLE.get(buf, 0).asInstanceOf[Double]
  }

  @throws[IOException]
  inline def writeDoubleRaw(os: OutputStream, value: Double): Unit = {
    val buf = new Array[Byte](lang.Double.BYTES)
    DOUBLE.set(buf, 0, value)
    os.write(buf)
  }

  infix def toHex(array: Array[Byte]): String = {
    if (array != null && !array.isEmpty) {
      val sb = new StringBuilder
      array.foreach(e => sb.addAll("%02X" format e))

      sb.toString
    } else {
      ""
    }
  }

  inline def hashCode(x: Array[Byte]): Int = {
    var result = 0

    if (x != null) {
      x.foreach(result + _)
    }

    result
  }
}
